Use `start`, `end`, and `order` query parameters to specify range in `listTransactions`.

jonathanknowles opened the pull request:
# Issue Number

#467 

# Overview

Implements the proposal outlined in https://github.com/input-output-hk/cardano-wallet/issues/467#issuecomment-514888695.

Improvements over the status quo:
* We no longer need to document and validate a special textual format for `Iso8601Range`.
* Dedicated handling of separators (`-`, and so on) is replaced with standardized parsing of query parameters (which is handled for us by Servant).
* Because we're no longer relying on the `-` separator to separate times, we can support all ISO 8601-formatted times, rather than just the basic format. We now simply reuse our existing parser, which is already tested.
* There's now a straightforward mapping from CLI arguments to API query parameters.
* There's now only one way to encode the infinite range: by not supplying either `start` or `end`. Previously, there were two ways to encode this range:
    * `Nothing`
    * `Just (Iso8601Range Nothing Nothing)`

This PR also:
* Adds an `order` parameter, which specifies a sort order for transactions returned by `listTransactions`.
* Upgrades our existing `transaction list` CLI integration test to also supply the `order` parameter.

# Work not tackled by this PR

- The proposal outlined in https://github.com/input-output-hk/cardano-wallet/issues/467#issuecomment-514888695 also suggests the use of a `limit` query parameter, to impose a limit on the number of transactions returned. This is left for a future PR.

- When filtering is actually implemented in the API server, we need to also make the API return a `400` error if `start > end`. We must also add CLI and API tests to cover this condition. This is left for a future PR.

jonathanknowles committed:
Move `Iso8601Time` type from `cli` to `core`.

jonathanknowles committed:
Use `start` and `end` query parameters instead of `Iso8601Range` type in API.

jonathanknowles committed:
Remove unused `Iso8601Range` type.

jonathanknowles committed:
Fix spelling mistake in function `httpApiDataRoundtrip`.

jonathanknowles committed:
Don't export `Iso8601Time` from `cli`.

jonathanknowles committed:
Add `SortOrder` enumeration type to represent a sort order, applicable to queries.

jonathanknowles committed:
Add `order` query parameter to the `ListTransactions` API call.

jonathanknowles committed:
Add `--order` parameter to `transaction list` CLI command.

jonathanknowles committed:
Save vertical space in `HttpApiData` instances for `Iso8601Time`.

KtorZ commented on lib/core/src/Cardano/Wallet/Api/Types.hs:
```diff
@@ -255,120 +259,49 @@ data ApiErrorCode
-            "*" -> pure Nothing
-            timeText -> pure <$> utcTimeFromText [iso8601BasicUtc] timeText
+instance FromText SortOrder where
+    fromText = fromTextToBoundedEnum CamelCase
```
mostly everything in our API is `snake_cased` isn't it? It won't matter much here since these are single-word constructors, but for consistency :grimacing: ...

jonathanknowles replied:
Fair enough! I'll update this.

jonathanknowles replied:
@KtorZ Fixed.

KtorZ commented on lib/core/test/unit/Cardano/Wallet/Api/TypesSpec.hs:
```diff
@@ -221,66 +222,12 @@ spec = do
-                    fromText @(Iso8601Range "test") text
-                        `shouldBe` Left (TextDecodingError $ err
-                            <> "Invalid end time string: "
-                            <> show invalidTime)
```
I think we still want some negative decoding paths for time strings, don't we ?

jonathanknowles replied:
> I think we still want some negative decoding paths for time strings, don't we ?

Indeed we do!

The `Iso8601Time` type reuses the functionality provided here: https://github.com/input-output-hk/cardano-wallet/blob/master/lib/core/src/Data/Time/Text.hs

We actually already have quite a few negative decoding examples here:
https://github.com/input-output-hk/cardano-wallet/blob/master/lib/core/test/unit/Data/Time/TextSpec.hs#L196-L271.

We also check that equivalent times with different local timezones are decoded equivalently:
https://github.com/input-output-hk/cardano-wallet/blob/master/lib/core/test/unit/Data/Time/TextSpec.hs#L57

Do you feel the coverage of this test suite is adequate? Are there any gaps you would like to fill?

Anviking commented on lib/cli/src/Cardano/CLI.hs:
```diff
@@ -570,6 +564,7 @@ data TransactionListArgs = TransactionListArgs
     , _walletId :: WalletId
     , _timeRangeStart :: Maybe Iso8601Time
     , _timeRangeEnd :: Maybe Iso8601Time
+    , _sortOrder :: Maybe SortOrder
```
Would `direction` be a more precise term?

My impression was that **order** refers to all possible `a -> a -> Ordering` (`Ordering = EQ, GT, LT`) implementations, whereas **direction** only specifies whether that function should be inverted or not.

I see [Heroku](https://devcenter.heroku.com/articles/platform-api-reference#ranges), as @KtorZ, mentioned uses `order` here. The [Github API](https://developer.github.com/v3/repos/#list-your-repositories) is an example that uses `direction`.

While on the topic, I think `since` and `until` as used in the [commits api](https://developer.github.com/v3/repos/commits/#list-commits-on-a-repository) are nice alternatives to `start` and `end`, but might not work well when used with `descending` order.

So I'm totally fine with the current naming if you deem it best.

jonathanknowles replied:
@Anviking wrote:

> Would `direction` be a more precise term?
> My impression was that **order** refers to all possible `a -> a -> Ordering` (`Ordering = EQ, GT, LT`) implementations, whereas **direction** only specifies whether that function should be inverted or not.

I think `SortDirection` could certainly also be acceptable.

However in that case, it would be probably be more natural to say "forward direction" and "reverse direction". (The terms "ascending direction" and "descending direction" sound quite unnatural to my ears.)

In my experience, "ascending order" and "descending order" are already commonly-used noun phrases, with well-defined meanings. For example, consider this usage (emphasis mine):

> "A **standard order** is often called **ascending** (corresponding to the fact that the standard order of numbers is ascending, i.e. A to Z, 0 to 9), the **reverse order** is often called **descending** (Z to A, 9 to 0). For dates and times, ascending means that earlier values precede later ones e.g. 1/1/2000 will sort ahead of 1/1/2001."
([source](https://en.wikipedia.org/wiki/Sorting))

Regarding:

> While on the topic, I think `since` and `until` as used in the [commits api](https://developer.github.com/v3/repos/commits/#list-commits-on-a-repository) are nice alternatives to `start` and `end`, but might not work well when used with `descending` order.

As I see it, the problem with the word "since" is that it has an _implied directionality_ that goes _forward in time_. But the words `start` and `end` simply define boundaries that could be anywhere in time. They could even be reversed, and it wouldn't contradict the English meanings. Consider this made-up example:

> "In order to produce the descending list of integers from 10 to 1, we **start** at 10, apply the `pred` function repeatedly, and stop when we reach the **end**, at 1".

If we try to substitute the word "since" here, then it doesn't work as nicely.

Anviking commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -461,9 +462,11 @@ listTransactions
+    -> Maybe SortOrder
     -> Handler [ApiTransaction t]
-listTransactions w (ApiT wid) _maybeRange = do
+listTransactions w (ApiT wid) _maybeStart _maybeEnd _maybeOrder = do
```
So… with this scheme, is it possible to have `end < start`?

What would `--start 2019-... --end 2000-...` do depending on `--order ascending` or `--order descending`?

jonathanknowles replied:
> So… with this scheme, is it possible to have `end < start`?
> What would `--start 2019-... --end 2000-...` do depending on `--order ascending` or `--order descending`?

This is an excellent point, and I'm glad you raised it.

We certainly have to clearly define the expected behaviour.

One way to handle this is to treat `start` and `end` as _directionless boundary markers_ when both are specified, and to always sort according to the `order` parameter (with a default of `ascending`, if no order is specified).

So for your example:
* `--start 2019 --end 2000`
      specifies all times between 2000 and 2019 in ascending order.
* `--start 2019 --end 2000 --order ascending`
      specifies all times between 2000 and 2019 in ascending order.
* `--start 2019 --end 2000 --order descending`
      specifies all times between 2000 and 2019 in descending order.

What do you think? Does it match your intuition?

jonathanknowles replied:
**Another way**, of course, is to use the **implied** order, if none is specified. So:

So for your example:
* `--start 2019 --end 2000`
      specifies all times between 2000 and 2019 in descending order.
* `--start 2019 --end 2000 --order ascending`
      specifies all times between 2000 and 2019 in ascending order.
* `--start 2019 --end 2000 --order descending`
      specifies all times between 2000 and 2019 in descending order.

jonathanknowles replied:
Whichever way we choose, our documentation and testing should be explicit and consistent.

jonathanknowles replied:
@Anviking I propose that we continue this discussion on the "Implement filtering on `listTransactions`" issue. I've added a comment here: https://github.com/input-output-hk/cardano-wallet/issues/466#issuecomment-515301573

We can implement the actual filtering with a future PR, which will depend on the outcome of the above discussion.

jonathanknowles replied:
@Anviking We've made the decision to throw a `400` error from the API if `start > end`. This will also need appropriate CLI and API tests. In order to keep the code review process simple, I propose tackling this in a further PR. What do you think? I've already added a work item to issue #466.

Anviking reviewed (commented):
Some naming-alternatives (sorry 😄), and a question about this new proposal.

jonathanknowles committed:
Use lower_snake_case instead of camelCase in textual representation of `SortOrder`.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/580#discussion_r307217092

Anviking commented on lib/core/src/Cardano/Wallet/Api/Types.hs:
```diff
@@ -255,120 +259,49 @@ data ApiErrorCode
-    = Iso8601Range (Maybe UTCTime) (Maybe UTCTime)
-    deriving (Eq, Generic, Show)
+newtype Iso8601Time = Iso8601Time
+    { getIso8601Time :: UTCTime
```
I noticed that `UTCTime` already conforms to `ToHttpApiData` and `FromHttpApiData` and works with the extended format, but not with the basic. As we're now not solely dependent on the basic format, we _could_ consider dropping support for it and the wrapper. But I wouldn't advocate it with all this already in place.

Anviking reviewed (approved):
lgtm!
